# Data Cleaning with Pandas


In [1]:
import pandas as pd

In [ ]:
# orders.csv
# url = "https://drive.google.com/file/d/1Vu0q91qZw6lqhIqbjoXYvYAQTmVHh6uZ/view?usp=sharing"
# path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
# orders = pd.read_csv(path)

# # orderlines.csv
# url = "https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=sharing"
# path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
# orderlines = pd.read_csv(path)

In [2]:
orders = pd.read_csv('3_Data_Cleaning/data/orders.csv', parse_dates=['created_date'])

In [3]:
orderlines = pd.read_csv('3_Data_Cleaning/data/orderlines.csv', parse_dates=['date'])

Before we begin, let's create a copy of the `orders` and `orderlines` DataFrames. This way we are sure any of our changes won't affect the original DataFrames

In [4]:
orders_df = orders.copy()

In [5]:
orderlines_df = orderlines.copy()

In [6]:
orders_df.sample(20)

order_id        created_date  total_paid            state
156937    457161 2017-12-07 15:12:24      158.99        Completed
132105    431747 2017-11-22 13:09:07        0.00  Shopping Basket
74606     374152 2017-07-07 11:50:47      329.99  Shopping Basket
152565    452768 2017-11-30 17:53:24     1123.00      Place Order
116610    416196 2017-10-23 12:18:56       62.79  Shopping Basket
26327     325807 2017-02-17 22:05:56      757.99  Shopping Basket
78807     378355 2017-07-17 10:01:11       74.90      Place Order
46727     346220 2017-04-14 21:52:32       63.99        Completed
203967    504455 2018-02-02 14:02:44        6.99      Place Order
162877    463164 2017-12-14 17:51:24      154.99      Place Order
196995    497479 2018-01-24 04:51:18     2850.57  Shopping Basket
64443     363988 2017-06-09 16:06:41      613.98  Shopping Basket
45667     345160 2017-04-10 18:22:24       97.99      Place Order
131704    431346 2017-11-22 09:36:09     3433.59  Shopping Basket
143382    443475 2017-11-25 15:15:19      224.12  Shopping Basket
18635     318115 2017-01-31 11:56:04      694.43      Place Order
47312     346805 2017-04-17 21:10:04      164.99        Completed
206303    506793 2018-02-06 12:53:23     1124.00  Shopping Basket
100992    400577 2017-09-15 22:10:34      352.45          Pending
174785    475173 2017-12-28 17:19:20       59.99      Place Order

In [7]:
orderlines_df.sample(20)

id  id_order  product_id  product_quantity      sku unit_price  \
33389   1190732    328723           0                 1  APP1669      39.99   
145611  1404413    425318           0                 1  APP1003   2.298.59   
168144  1447683    443617           0                 1  LGE0037     140.06   
237255  1559656    490512           0                 1  SEA0099      99.00   
91258   1293841    377927           0                 1  STA0021       6.99   
162068  1437636    439308           0                 1  APP1922     154.00   
188217  1477230    455458           0                 1  PAC2114   3.019.59   
225976  1540730    483297           0                 1  APP1937   1.278.59   
106573  1321782    391264           0                 1  APP1966     360.33   
224793  1538886    482504           0                 1  APP0663     199.00   
258595  1594034    503468           0                 1  PAC1942   4.074.59   
18911   1162365    316206           0                 1  TRA0034      29.90   
8080    1140183    306934           0                 1  ELE0007      26.90   
22575   1170273    319442           0                 1  THU0013      75.99   
150559  1417665    429856           0                 1  MUJ0033      32.29   
32514   1189022    327932           0                 1  MAC0121       7.99   
132924  1381536    414231           0                 1  CRU0050      91.58   
253891  1586115    500672           0                 1  MIC0036      64.99   
281409  1631182    519281           0                 1  APP2486   1.067.00   
108807  1325780    393223           0                 1  SSE0007      25.99   

                      date  
33389  2017-02-24 21:41:15  
145611 2017-11-11 15:47:46  
168144 2017-11-25 17:17:06  
237255 2018-01-14 14:07:36  
91258  2017-07-15 20:46:38  
162068 2017-11-24 14:12:33  
188217 2017-12-05 05:15:07  
225976 2018-01-06 03:12:24  
106573 2017-08-21 15:55:58  
224793 2018-01-04 21:59:06  
258595 2018-02-01 00:32:11  
18911  2017-01-28 07:27:54  
8080   2017-01-10 23:27:12  
22575  2017-02-03 08:41:20  
150559 2017-11-21 01:05:29  
32514  2017-02-23 10:14:37  
132924 2017-10-18 23:07:40  
253891 2018-01-29 06:04:20  
281409 2018-03-01 09:07:23  
108807 2017-08-28 00:24:18

## 1.&nbsp; Duplicates
We can check for duplicates using the pandas [.duplicated()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html) method.

We can then delete these rows, if we wish, using [.drop_duplicates()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html)

## 1.1 Orders

In [8]:
# orders_df
orders_df.duplicated().sum()

np.int64(0)

In [9]:
# Checking for duplicates in the key column (('order_id'))
orders_df.duplicated(subset='order_id').sum()

np.int64(0)

No duplicated rows or keys found in the "orders" dataframe.

## 1.2 Orderlines

In [10]:
# orderlines_df
orderlines_df.duplicated().sum()

np.int64(0)

In [11]:
# Checking for duplicates in the key column ('id')
orderlines_df.duplicated(subset='id').sum()

np.int64(0)

In [12]:
# Checking for duplicates in the key column
orderlines_df.duplicated(subset=['id', 'id_order']).sum()

np.int64(0)

No duplicated rows or keys found in the "orderlines" dataframe.

In [13]:
# Checking for duplicates in the foreign key column ('id_order')
orderlines_df.duplicated(subset=['id_order']).sum()

np.int64(89128)

In [14]:
orderlines_df[orderlines_df.duplicated(subset=['id_order'], keep=False)]

id  id_order  product_id  product_quantity      sku unit_price  \
13      1119129    299552           0                 1  MBI0010      16.99   
14      1119130    299552           0                 1  SAT0044      19.99   
21      1119140    299559           0                 1  APP0700      72.19   
22      1119143    299559           0                 1  APP1214      84.99   
33      1119181    299571           0                 1  SYN0139     166.24   
...         ...       ...         ...               ...      ...        ...   
293954  1650162    527369           0                 1  BEL0307      27.99   
293956  1650166    527377           0                 1  APP0698       9.99   
293957  1650168    527377           0                 1  APP2501     391.00   
293966  1650183    527388           0                 1  TPL0025      14.99   
293981  1650202    527388           0                 1  BEZ0204      19.99   

                      date  
13     2017-01-01 02:13:51  
14     2017-01-01 02:14:36  
21     2017-01-01 02:32:22  
22     2017-01-01 02:36:43  
33     2017-01-01 04:00:44  
...                    ...  
293954 2018-03-14 13:42:41  
293956 2018-03-14 13:43:47  
293957 2018-03-14 13:44:27  
293966 2018-03-14 13:51:59  
293981 2018-03-14 13:58:01  

[137842 rows x 7 columns]

There is a **one to many** relationship between the tables orders and orderlines. The foreign key 'id_order' in the table orderlines is linked to the primary key 'order_id' in the table orders.

# 2.&nbsp; `.info()`

In [15]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      226909 non-null  int64         
 1   created_date  226909 non-null  datetime64[ns]
 2   total_paid    226904 non-null  float64       
 3   state         226909 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 6.9+ MB


* `total_paid` has 5 missing values
* `created_date` should become datetime datatype --- DONE! Imported as "datetime" datatype.

In [16]:
orders_df.nunique()

order_id        226909
created_date    224828
total_paid       31236
state                5
dtype: int64

In [17]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 non-null  int64         
 4   sku               293983 non-null  object        
 5   unit_price        293983 non-null  object        
 6   date              293983 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(2)
memory usage: 15.7+ MB


* `date` should be a datetime datatype  --- DONE! Imported as "datetime" datatype.
* `unit_price` should be a float datatype

In [18]:
orderlines_df.nunique()

id                  293983
id_order            204855
product_id               1
product_quantity        67
sku                   7951
unit_price           11329
date                251631
dtype: int64

## 3.&nbsp; Missing values

### 3.1.&nbsp; Orders
* `total_paid` has 5 missing values

In [19]:
num_missing = orders_df['total_paid'].isna().sum()
total_rows = orders_df.shape[0]
percent_missing = (100*num_missing/total_rows)
print(f"{num_missing} missing values represents {percent_missing:.5f}% of the {total_rows} rows in our DataFrame")

5 missing values represents 0.00220% of the 226909 rows in our DataFrame


> A quick way to find out a percentage if you don't need to print out a sentence for yourself/colleagues is `.value_count(normalize=True)`

In [20]:
orders_df['total_paid'].isna().value_counts(normalize=True)

total_paid
False    0.999978
True     0.000022
Name: proportion, dtype: float64

### Imputing 'total_paid' missing values

As the number (percentage) of missing values is very small, we can just delete these rows from our DataFrame.
In addition, we will delete the corresponding order_id's in the orderlines dataframe.

In [21]:
# order_id's of 'total_paid' missing values
order_id_missing_df = orders_df.loc[orders_df['total_paid'].isna(),'order_id']
order_id_missing_df

127701    427314
132013    431655
147316    447411
148833    448966
149434    449596
Name: order_id, dtype: int64

In [22]:
# Dropping nan from orders_df
orders_df = orders_df.dropna(axis=0)
orders_df

order_id        created_date  total_paid            state
0         241319 2017-01-02 13:35:40       44.99        Cancelled
1         241423 2017-11-06 13:10:02      136.15        Completed
2         242832 2017-12-31 17:40:03       15.76        Completed
3         243330 2017-02-16 10:59:38       84.98        Completed
4         243784 2017-11-24 13:35:19      157.86        Cancelled
...          ...                 ...         ...              ...
226904    527397 2018-03-14 13:56:38       42.99      Place Order
226905    527398 2018-03-14 13:57:25       42.99  Shopping Basket
226906    527399 2018-03-14 13:57:34      141.58  Shopping Basket
226907    527400 2018-03-14 13:57:41       19.98  Shopping Basket
226908    527401 2018-03-14 13:58:36       18.98      Place Order

[226904 rows x 4 columns]

In [23]:
mask = orderlines_df['id_order'].isin(order_id_missing_df)

In [24]:
# Corresponding 'id_order' in orderlines table
orderlines_df.loc[mask,:]

id  id_order  product_id  product_quantity      sku unit_price  \
150248  1416957    427314           0                 1  MUJ0023      26.99   
150252  1416966    427314           0                 1  APP1696      49.00   
152847  1421402    431655           0                 1  APP1922     154.00   
173200  1456182    447411           0                 1  LAC0225     180.76   
175213  1459674    448966           0                 1  PAC2266     961.02   
176040  1461236    449596           0                 1  APP2191   1.722.59   

                      date  
150248 2017-11-20 18:40:45  
150252 2017-11-20 18:44:30  
152847 2017-11-22 12:12:16  
173200 2017-11-27 10:30:28  
175213 2017-11-27 18:03:18  
176040 2017-11-27 21:49:21

In [25]:
# Before dropping the rows, we will save the dropped id_orders to a new dataframe
dropped_ids_missing_orders_total_paid_df = pd.DataFrame(orderlines_df.loc[mask, 'id_order']).copy().reset_index(drop=True)
dropped_ids_missing_orders_total_paid_df['reason'] = 'Missing column "total_paid" in orders_df dataset'
dropped_ids_missing_orders_total_paid_df

id_order                                            reason
0    427314  Missing column "total_paid" in orders_df dataset
1    427314  Missing column "total_paid" in orders_df dataset
2    431655  Missing column "total_paid" in orders_df dataset
3    447411  Missing column "total_paid" in orders_df dataset
4    448966  Missing column "total_paid" in orders_df dataset
5    449596  Missing column "total_paid" in orders_df dataset

In [26]:
# Drop the rows from the orderlines_df
orderlines_df = orderlines_df[~mask].reset_index(drop=True)
orderlines_df

id  id_order  product_id  product_quantity      sku unit_price  \
0       1119109    299539           0                 1  OTT0133      18.99   
1       1119110    299540           0                 1  LGE0043     399.00   
2       1119111    299541           0                 1  PAR0071     474.05   
3       1119112    299542           0                 1  WDT0315      68.39   
4       1119113    299543           0                 1  JBL0104      23.74   
...         ...       ...         ...               ...      ...        ...   
293972  1650199    527398           0                 1  JBL0122      42.99   
293973  1650200    527399           0                 1  PAC0653     141.58   
293974  1650201    527400           0                 2  APP0698       9.99   
293975  1650202    527388           0                 1  BEZ0204      19.99   
293976  1650203    527401           0                 1  APP0927      13.99   

                      date  
0      2017-01-01 00:07:19  
1      2017-01-01 00:19:45  
2      2017-01-01 00:20:57  
3      2017-01-01 00:51:40  
4      2017-01-01 01:06:38  
...                    ...  
293972 2018-03-14 13:57:25  
293973 2018-03-14 13:57:34  
293974 2018-03-14 13:57:41  
293975 2018-03-14 13:58:01  
293976 2018-03-14 13:58:36  

[293977 rows x 7 columns]

In [27]:
# Checking for missing values
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 226904 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      226904 non-null  int64         
 1   created_date  226904 non-null  datetime64[ns]
 2   total_paid    226904 non-null  float64       
 3   state         226904 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 8.7+ MB


### 3.2.&nbsp; Orderlines
There are no missing values in `orderlines_df`

## 4.&nbsp; Datatypes

### 4.1.&nbsp; Orders
* `created_date` should become datetime datatype. Type conversion was done during csv file import.

In [ ]:
# orders_df["created_date"] = pd.to_datetime(orders_df["created_date"])

### 4.1.&nbsp; Orderlines
* `date` should be a datetime datatype. Type conversion was done during csv file import.
* `unit_price` should be a float datatype

#### 4.1.1.&nbsp; `date`

In [ ]:
# orderlines_df["date"] = pd.to_datetime(orderlines_df["date"])

#### 4.1.2.&nbsp;`unit_price`

In [28]:
orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"])

ValueError: Unable to parse string "1.137.99" at position 6

As you can see when we try to convert `unit_price` to a numerical datatype, we receive a `ValueError` telling us that pandas doesn't understand the number `1.137.99`. This is probably because numbers cannot have multiple decimal points. Let's see if there are any other numbers like this:

> `.` is a wildcard in regex, we need the `\` as an escape

In [29]:
# Count the number of decimal points in the unit_price
orderlines_df['unit_price'].str.count("\\.").value_counts()

unit_price
1    257809
2     36168
Name: count, dtype: int64

Looks like over 36000 rows in `orderlines` are affected by this problem. Let's work out how much that is as a percentage of our total data.

In [30]:
# Count the rows with more than one `.`
mult_decimal_rows = (orderlines_df['unit_price'].str.count("\\.")>1).sum()

# Find the percentage of corrupted rows
percent_corrupted = (100 * mult_decimal_rows / orderlines_df.shape[0])
print(f"{percent_corrupted:.2f}% of the rows in our DataFrame have multiple decimal points in the unit_price")

12.30% of the rows in our DataFrame have multiple decimal points in the unit_price


This is a bit of a tricky decision as 12.3% is a significant amount of our data... and we might even end up losing a larger portion of our data than this too. For the moment we will delete the rows as we only have 2 weeks for this project and I'd like some quick, accurate results to show. If we have time at the end, we can come back and investigate this problem further, maybe there's a solution?

Each row of `orderlines` represents a product in an order. For example, if order number 175 contained 3 seperate products, then order 175 would have 3 rows in `orderlines`, one row for each of the products. If 2 of those products have 'normal' prices (14.99, 15.85) and 1 has a price with 2 decimal points (1.137.99), we need to remove the whole order and not just the affected row. If we only remove the row with 2 decimal places then any later analysis about products and prices could be misleading.

We therefore need to find the order numbers associated with the rows that have 2 decimal points, and then remove all the associated rows.

In [31]:
# Boolean mask to find the orders that contain a price with multiple decimal points
multiple_decimal_mask = orderlines_df['unit_price'].str.count("\\.") > 1

# Apply the boolean mask to the orderlines DataFrame. This way we can find the order_id of all the affected orders.
corrupted_order_ids = orderlines_df.loc[multiple_decimal_mask, "id_order"]

# Keep only the rows that do not have multiple decimal points
#orderlines_df = orderlines_df.loc[~orderlines_df['id_order'].isin(corrupted_order_ids)]

#### Correcting unit_price data type

In [32]:
orderlines_df.loc[multiple_decimal_mask, 'unit_price'] = orderlines_df.loc[multiple_decimal_mask, 'unit_price'].str.replace(".", "", 1)

# Converting unit_price type to numeric
orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"])

orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293977 entries, 0 to 293976
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293977 non-null  int64         
 1   id_order          293977 non-null  int64         
 2   product_id        293977 non-null  int64         
 3   product_quantity  293977 non-null  int64         
 4   sku               293977 non-null  object        
 5   unit_price        293977 non-null  float64       
 6   date              293977 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(4), object(1)
memory usage: 15.7+ MB


It worked perfectly

# Clean the `products` DataFrame

In [ ]:
# # products.csv
# url = "https://drive.google.com/file/d/1afxwDXfl-7cQ_qLwyDitfcCx3u7WMvkU/view?usp=sharing"
# path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
# products = pd.read_csv(path)

In [33]:
products = pd.read_csv('3_Data_Cleaning/data/products.csv')

In [34]:
products_df = products.copy()

In [35]:
products_df.sample(20)

sku                                               name  \
19027    PAC2503  OWC ThuderBay RAID Thunderbolt 4 3 | 24TB (4x6...   
14360    APP1825  Apple MacBook Pro 13 "Core i7 24GHz | 8GB RAM ...   
10212    PAC1594  Apple iMac 27 "Core i5 3.2GHz Retina 5K | 32GB...   
11542    TWS0095  Twelve South BookBook MacBook sleeve Rutledge ...   
10154    PAC1592  Apple iMac 27 "Core i5 3.2GHz Retina 5K | 8GB ...   
2684     QNA0116       QNAP TVS-863 | 4GB RAM Mac and PC Server Nas   
18922  LIF0114-A  Open - LifeProof nüüd Submersible iPhone Case ...   
16120    AKI0019   Akitio Thunder2 QUAD Case External Thunderbolt 2   
6633     PAC1238  Apple iMac 27 "Core i7 Retina 5K 4GHz | 32GB |...   
6971     PAC1610  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 8GB ...   
3018     ZAG0021  Zagg invisibleSHIELD Screen Protector Apple Wa...   
8306     PAC1585  Apple iMac 21.5 "Core i7 3.3GHz 4K Retina | 16...   
5065     PAC1045  Apple iMac 27 "Core i7 Retina 5K 4GHz | 16GB |...   
12283    APP1468   Apple Watch Sport Silver 38mm white sports strap   
4828     PAC1041  Apple iMac 27 "Core i7 Retina 5K 4GHz | 16GB |...   
15390    REP0300                   Full screen repair iPhone 6 Plus   
912      GRT0312  Griffin Multidock February 10 bays Cargo Secur...   
6650     PAC1237  Apple iMac 27 "Core i7 Retina 5K 4GHz | 16GB |...   
13459    MOS0205             Moshi iGlaze iPhone Armor Case 7 Black   
17558    AP20332  Like new - Apple Smart Cover iPad Mini 4 Charc...   

                                                    desc    price promo_price  \
19027  RAID Disk external box 25 inch and 35 inch two...  1724.95  13.413.685   
14360  MacBook Pro 13 inch i7 to 24GHz with 8GB RAM a...     2299  21.055.936   
10212  Desktop computer iMac 27-inch 3.2GHz Core i5 5...     3889  28.489.898   
11542  Cover hardcover book design and vintage leathe...    99.99     929.885   
10154  Desktop computer iMac 27-inch 3.2GHz Core i5 5...     3169  25.979.898   
2684                    8-bay NAS Server for Mac and PC.  1136.19  11.359.904   
18922  submersible and resistant to extreme condition...    89.99      42.729   
16120  Disk outer box 25 inch and 35 inch Thunderbolt...      468   4.315.114   
6633   IMac desktop computer 27 inch Retina 5K RAM 32...     4669  37.695.036   
6971   Desktop computer iMac 27-inch 3.3GHz Core i5 5...     4069  31.889.901   
3018                   Screen Protector for Apple Watch.     9.99      69.902   
8306   Desktop computer iMac 21.5-inch 3.3GHz Core i7...     3649     2749.99   
5065   IMac desktop computer 27 inch 5K Retina i7 4GH...     3409  31.019.899   
12283  Apple Watch Sport Silver 38mm white sports bel...      369   3.690.004   
4828   IMac desktop computer 27 inch 5K Retina i7 4GH...     3169  28.739.896   
15390  Repair service including parts and labor for i...    99.99     999.899   
912                         charging station 10 devices.      899   8.989.937   
6650   IMac desktop computer 27 inch Retina 5K RAM 16...     4189  35.949.899   
13459  protective cover two colors with metallic acce...       40     329.894   
17558  Polyurethane Smart Cover for iPad Reconditione...       45     219.905   

       in_stock      type  
19027         0  11935397  
14360         0      2158  
10212         0  5,74E+15  
11542         0  13835403  
10154         0  5,74E+15  
2684          0  12175397  
18922         1  11865403  
16120         0  12995397  
6633          0  5,74E+15  
6971          0  5,74E+15  
3018          1      2425  
8306          0      1282  
5065          0  5,74E+15  
12283         0  24885185  
4828          0  5,74E+15  
15390         0  1,44E+11  
912           0      1216  
6650          0  5,74E+15  
13459         0  11865403  
17558         0  12635403

# 1. Look for Duplicates

In [36]:
products_df.duplicated().sum()

np.int64(8746)

There are duplicated rows in the "products" DataFrame.

In [37]:
# Checking for duplicates in the key column ('sku')
products_df[products_df.duplicated(subset=['sku'], keep=False)]

sku                                               name  \
100    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
101    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
102    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
103    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
104    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
...        ...                                                ...   
18190  PAR0077                         Parrot Bebop Drone 2 Power   
18307  NKI0010              Nokia Wireless sphygmomanometer Plata   
18308  NKI0010              Nokia Wireless sphygmomanometer Plata   
18513  PLA0035          Bluetooth Headset Plantronics Explorer 80   
18514  PLA0035          Bluetooth Headset Plantronics Explorer 80   

                                                    desc   price promo_price  \
100          MacBook Pro laptop 133 inches (MD101Y / A).    1199  11.455.917   
101          MacBook Pro laptop 133 inches (MD101Y / A).    1199  11.455.917   
102          MacBook Pro laptop 133 inches (MD101Y / A).    1199  11.455.917   
103          MacBook Pro laptop 133 inches (MD101Y / A).    1199  11.455.917   
104          MacBook Pro laptop 133 inches (MD101Y / A).    1199  11.455.917   
...                                                  ...     ...         ...   
18190  Drone cuadricóptero quality camera integrated ...   699.9   6.733.892   
18307     Sphygmomanometer for iPhone iPad and iPod App.  129.99   1.149.899   
18308     Sphygmomanometer for iPhone iPad and iPod App.  129.99   1.149.899   
18513  egonómico wireless headset design for iPhone i...   39.99     349.896   
18514  egonómico wireless headset design for iPhone i...   39.99     349.896   

       in_stock      type  
100           0      1282  
101           0      1282  
102           0      1282  
103           0      1282  
104           0      1282  
...         ...       ...  
18190         0  11905404  
18307         1  11905404  
18308         1  11905404  
18513         1      5384  
18514         1      5384  

[9504 rows x 7 columns]

In [38]:
num_duplicates = products_df.duplicated(subset='sku').sum()
total_rows = products_df.shape[0]
percent_duplicated = (100*num_duplicates/total_rows)
print(f"{num_duplicates} duplicated rows represents {percent_duplicated:.5f}% of the {total_rows} rows in our DataFrame")

8747 duplicated rows represents 45.26027% of the 19326 rows in our DataFrame


In [39]:
# Drop duplicates
products_df = products_df.drop_duplicates()

In [40]:
# Checking for duplicates in the key column ('sku')
products_df.duplicated(subset='sku').sum()

np.int64(1)

There is still one duplicate value in the SKU (key) column.

In [41]:
# Checking for duplicates in the key column ('sku')
products_df[products_df.duplicated(subset=['sku'], keep=False)]

sku                                               name  \
7992  APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   
8000  APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   

                                                   desc price promo_price  \
7992  Desktop Apple iMac 21.5 inch i5 31 GHz Retina ...  1729     1305.59   
8000  Desktop Apple iMac 21.5 inch i5 31 GHz Retina ...   NaN     1305.59   

      in_stock  type  
7992         0  1282  
8000         0  1282

In [42]:
# Drop the duplicated entry, we keep the first row of the duplicate group
products_df = products_df.drop_duplicates(subset='sku', keep='first')

In [43]:
# Checking for duplicates in the key column ('sku')
products_df.duplicated(subset='sku').sum()

np.int64(0)

# 2. Look for Missing values


In [44]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10579 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10579 non-null  object
 1   name         10579 non-null  object
 2   desc         10572 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10579 non-null  object
 5   in_stock     10579 non-null  int64 
 6   type         10529 non-null  object
dtypes: int64(1), object(6)
memory usage: 661.2+ KB


* `desc` has 7 missing values
* `price` has 45 missing values
* `type` has 50 missing values
  
* `price` should become a float datatype.
* `promo_price` should become a float datatype

In [45]:
# SKU's of 'price' missing values
sku_missing_df = products_df.loc[products_df['price'].isna(),'sku']
sku_missing_df

34         TWS0019
1900       AII0008
2039       CEL0020
2042       CEL0007
2043       CEL0012
2044       CEL0014
2049       CEL0015
2051       CEL0018
2052       CEL0023
2053       CEL0025
12052      CEL0028
12053      CEL0029
12054      CEL0030
12055      CEL0031
12059      CEL0032
12060      CEL0042
12061      CEL0043
12062      CEL0044
12064      CEL0038
12067      CEL0033
12068      CEL0034
12069      CEL0035
12070      CEL0036
12071      CEL0051
12072      CEL0052
12073      CEL0053
12074      CEL0054
12075      CEL0055
12076      CEL0056
12077      CEL0057
12078      CEL0058
12079      CEL0059
12080      CEL0060
12088      CEL0040
12089      CEL0041
12092      CEL0045
12093      CEL0046
12095      CEL0047
12096      CEL0048
12097      CEL0049
12098      CEL0050
12100      CEL0037
12238      CEL0027
12242      CEL0039
12947    CEL0034-A
Name: sku, dtype: object

In [46]:
mask = orderlines_df['sku'].isin(sku_missing_df)

In [47]:
# Corresponding 'sku' in orderlines table
orderlines_df.loc[mask,:]

id  id_order  product_id  product_quantity      sku  unit_price  \
1404    1122003    300843           0                 1  CEL0044        9.99   
3646    1130654    302872           0                 1  CEL0028        9.99   
4049    1131495    303274           0                 2  CEL0046       14.99   
4144    1131746    303363           0                 1  CEL0036        8.99   
4362    1132171    303577           0                 1  TWS0019       29.99   
...         ...       ...         ...               ...      ...         ...   
244780  1571962    495201           0                 1  CEL0028        6.99   
252599  1584351    500011           0                 1  CEL0028        6.99   
266957  1606736    508491           0                 1  CEL0028        5.99   
283897  1635382    521165           0                 1  CEL0044        7.99   
285005  1636732    521667           0                 1  CEL0044        7.99   

                      date  
1404   2017-01-02 18:45:40  
3646   2017-01-05 01:44:23  
4049   2017-01-05 16:47:42  
4144   2017-01-05 20:38:40  
4362   2017-01-06 12:33:30  
...                    ...  
244780 2018-01-20 19:19:10  
252599 2018-01-28 09:27:04  
266957 2018-02-09 00:24:58  
283897 2018-03-05 07:35:33  
285005 2018-03-05 13:39:35  

[124 rows x 7 columns]

124 orders are affected by prices with missing values (products_df). We will drop these orders to avoid missing values in our analysis. We will also save the dropped id_orders to a new dataframe for reference.

In [48]:
# Before dropping the rows, we will save the dropped id_orders to a new dataframe
dropped_ids_missing_sku_prices_df = pd.DataFrame(orderlines_df.loc[mask, 'id_order']).copy().reset_index(drop=True)
dropped_ids_missing_sku_prices_df['reason'] = "Missing column 'price' in products_df dataset"
dropped_ids_missing_sku_prices_df

id_order                                         reason
0      300843  Missing column 'price' in products_df dataset
1      302872  Missing column 'price' in products_df dataset
2      303274  Missing column 'price' in products_df dataset
3      303363  Missing column 'price' in products_df dataset
4      303577  Missing column 'price' in products_df dataset
..        ...                                            ...
119    495201  Missing column 'price' in products_df dataset
120    500011  Missing column 'price' in products_df dataset
121    508491  Missing column 'price' in products_df dataset
122    521165  Missing column 'price' in products_df dataset
123    521667  Missing column 'price' in products_df dataset

[124 rows x 2 columns]

In [49]:
# Drop the rows from the orderlines_df
orderlines_df = orderlines_df[~mask].reset_index(drop=True)
orderlines_df

id  id_order  product_id  product_quantity      sku  unit_price  \
0       1119109    299539           0                 1  OTT0133       18.99   
1       1119110    299540           0                 1  LGE0043      399.00   
2       1119111    299541           0                 1  PAR0071      474.05   
3       1119112    299542           0                 1  WDT0315       68.39   
4       1119113    299543           0                 1  JBL0104       23.74   
...         ...       ...         ...               ...      ...         ...   
293848  1650199    527398           0                 1  JBL0122       42.99   
293849  1650200    527399           0                 1  PAC0653      141.58   
293850  1650201    527400           0                 2  APP0698        9.99   
293851  1650202    527388           0                 1  BEZ0204       19.99   
293852  1650203    527401           0                 1  APP0927       13.99   

                      date  
0      2017-01-01 00:07:19  
1      2017-01-01 00:19:45  
2      2017-01-01 00:20:57  
3      2017-01-01 00:51:40  
4      2017-01-01 01:06:38  
...                    ...  
293848 2018-03-14 13:57:25  
293849 2018-03-14 13:57:34  
293850 2018-03-14 13:57:41  
293851 2018-03-14 13:58:01  
293852 2018-03-14 13:58:36  

[293853 rows x 7 columns]

In [50]:
# Dropping rows with missing values in the 'price' column
products_df = products_df.dropna(subset=['price'])

In [51]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10534 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10534 non-null  object
 1   name         10534 non-null  object
 2   desc         10527 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10534 non-null  object
 5   in_stock     10534 non-null  int64 
 6   type         10484 non-null  object
dtypes: int64(1), object(6)
memory usage: 658.4+ KB


No more missing values in the 'price' column. Since the columns 'desc' and 'type' have low relevance for the current analysis, we leave the missing values for these columns in the dataset.

# 3. Check / Change Data types

### 3.1 &nbsp;`price`

* `price` should be a float datatype

In [52]:
# This does not work because format problems with the price
products_df["price"] = pd.to_numeric(products_df["price"])

ValueError: Unable to parse string "1.639.792" at position 504

As you can see when we try to convert `price` to a numerical datatype, we receive a `ValueError` telling us that pandas doesn't understand the number `1.639.792`. This is probably because numbers cannot have multiple decimal points. Let's see if there are any other numbers like this:

In [53]:
# Count the number of decimal points in the unit_price
products_df['price'].str.count("\\.").value_counts()

price
1    6942
0    3215
2     377
Name: count, dtype: int64

Prices with zero decimal points are fine for the analysis.

However, 377 rows in `products` have 2 or more decimal points. Let's work out how much that is as a percentage of our total data.

In [54]:
# Count the rows with more than one `.`
mult_decimal_rows = (products_df['price'].str.count("\\.")>1).sum()

# Find the percentage of corrupted rows
percent_corrupted = (100 * mult_decimal_rows / products_df.shape[0])
print(f"{percent_corrupted:.2f}% of the rows in our DataFrame have multiple decimal points in the price")

3.58% of the rows in our DataFrame have multiple decimal points in the price


### Handling of products with two dots in the price
Let's see if all prices that look like this have been corrupted with a rule that we can reverse-engineer, and fix the prices.

In [55]:
# Boolean mask to find the orders that contain a price with multiple decimal points
multiple_decimal_mask = products_df['price'].str.count("\\.") > 1

# Apply the boolean mask to the products DataFrame. This way we can find the sku's of all the affected products.
corrupted_sku = products_df.loc[multiple_decimal_mask, ["sku",'price']]

In [56]:
# Dataframe with sku's that have more than 1 decimal point
corrupted_sku

sku      price
665    CRU0015-2  1.639.792
792      APP0672  4.694.994
797      APP0673  4.090.042
827      PAC0339  2.199.791
885      PAC0376  5.609.698
...          ...        ...
19312    REP0424  6.999.003
19313    REP0421  6.999.003
19314    REP0416  6.999.003
19315    REP0413  6.999.003
19316    REP0407  6.999.003

[377 rows x 2 columns]

In [57]:
mask = orderlines_df['sku'].isin(corrupted_sku['sku'])

In [58]:
# Corresponding 'sku' in orderlines table
orderlines_df.loc[mask,:]

id  id_order  product_id  product_quantity      sku  unit_price  \
341     1119745    299860           0                 1  PAC1516      457.89   
387     1119837    299902           0                 1  IOT0018       23.74   
706     1120440    300202           0                 1  KIN0150       10.99   
1103    1121367    300564           0                 1  APP0870      519.99   
1162    1121501    300621           0                 1  IOT0018       24.99   
...         ...       ...         ...               ...      ...         ...   
293279  1649114    526844           0                 1  APP2494      219.00   
293518  1649570    527099           0                 2  CAL0010      299.99   
293708  1649954    527286           0                 1  APP2490     1159.00   
293781  1650088    527342           0                 1  APP2492     1329.00   
293806  1650135    527364           0                 1  APP2491     1329.00   

                      date  
341    2017-01-01 18:40:38  
387    2017-01-01 19:32:46  
706    2017-01-02 01:31:40  
1103   2017-01-02 13:50:49  
1162   2017-01-02 15:02:38  
...                    ...  
293279 2018-03-14 10:43:52  
293518 2018-03-14 11:54:58  
293708 2018-03-14 12:55:14  
293781 2018-03-14 13:24:51  
293806 2018-03-14 13:36:13  

[4150 rows x 7 columns]

4150 orders are affected by prices with 2 or more decimal points (products_df). We will drop these orders to avoid erroneous prices in our analysis. We will also save the dropped id_orders to a new dataframe for reference.

In [59]:
# Before dropping the rows, we will save the dropped id_orders to a new dataframe
dropped_ids_mult_dots_sku_prices_df = pd.DataFrame(orderlines_df.loc[mask, 'id_order']).copy().reset_index(drop=True)
dropped_ids_mult_dots_sku_prices_df['reason'] = "Multiple decimal points in column 'price' in products_df dataset"
dropped_ids_mult_dots_sku_prices_df

id_order                                             reason
0       299860  Multiple decimal points in column 'price' in p...
1       299902  Multiple decimal points in column 'price' in p...
2       300202  Multiple decimal points in column 'price' in p...
3       300564  Multiple decimal points in column 'price' in p...
4       300621  Multiple decimal points in column 'price' in p...
...        ...                                                ...
4145    526844  Multiple decimal points in column 'price' in p...
4146    527099  Multiple decimal points in column 'price' in p...
4147    527286  Multiple decimal points in column 'price' in p...
4148    527342  Multiple decimal points in column 'price' in p...
4149    527364  Multiple decimal points in column 'price' in p...

[4150 rows x 2 columns]

In [60]:
# Drop the rows from the orderlines_df
orderlines_df = orderlines_df[~mask].reset_index(drop=True)
orderlines_df

id  id_order  product_id  product_quantity      sku  unit_price  \
0       1119109    299539           0                 1  OTT0133       18.99   
1       1119110    299540           0                 1  LGE0043      399.00   
2       1119111    299541           0                 1  PAR0071      474.05   
3       1119112    299542           0                 1  WDT0315       68.39   
4       1119113    299543           0                 1  JBL0104       23.74   
...         ...       ...         ...               ...      ...         ...   
289698  1650199    527398           0                 1  JBL0122       42.99   
289699  1650200    527399           0                 1  PAC0653      141.58   
289700  1650201    527400           0                 2  APP0698        9.99   
289701  1650202    527388           0                 1  BEZ0204       19.99   
289702  1650203    527401           0                 1  APP0927       13.99   

                      date  
0      2017-01-01 00:07:19  
1      2017-01-01 00:19:45  
2      2017-01-01 00:20:57  
3      2017-01-01 00:51:40  
4      2017-01-01 01:06:38  
...                    ...  
289698 2018-03-14 13:57:25  
289699 2018-03-14 13:57:34  
289700 2018-03-14 13:57:41  
289701 2018-03-14 13:58:01  
289702 2018-03-14 13:58:36  

[289703 rows x 7 columns]

In [61]:
# Dropping rows with multiple dots in the 'price' column
products_df = products_df[~multiple_decimal_mask].reset_index(drop=True)
products_df

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
10152  BEL0376        Belkin Travel Support Apple Watch Black   
10153  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
10154  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
10155  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
10156  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc  price promo_price  \
0           Aluminum support compatible with all MacBook  59.99     499.899   
1              USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                                 mouse Apple USB cable.     59     569.898   
3                  IPhone dock and USB Cable Apple iPod.     25     229.997   
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
...                                                  ...    ...         ...   
10152  compact and portable stand vertically or horiz...  29.99     269.903   
10153  Backpack with capacity of 14 liter compartment...  69.95     649.903   
10154  Backpack with capacity of 14 liter compartment...  69.95     649.903   
10155  Backpack with capacity of 14 liter compartment...  69.95     649.903   
10156  Backpack with capacity of 14 liter compartment...  69.95     649.903   

       in_stock      type  
0             1      8696  
1             0  13855401  
2             0      1387  
3             0      1230  
4             1      1364  
...         ...       ...  
10152         1     12282  
10153         1      1392  
10154         1      1392  
10155         0      1392  
10156         1      1392  

[10157 rows x 7 columns]

In [62]:
# CHECK: Count the number of decimal points in the price
products_df['price'].str.count("\\.").value_counts()

price
1    6942
0    3215
Name: count, dtype: int64

### Checking of prices with more than 2 decimal numbers (and only one decimal mark)

Some prices have only one decimal mark, but more than two decimal places (e.g. 99.004). Some of these are erroneous and can result in incorrect discount calculations. We will drop them from our analysis.

In [63]:
products_df.sample(20)

sku                                               name  \
3418     QNA0156       QNAP TS-653A | 4GB RAM Mac and PC Server NAS   
153      OWC0024    OWC Mount Adapter 25-35 Pro Mac Pro (2009-2012)   
6294   SAT0002-A  Open - Satechi USB Hub Slim-C to USB-A / HDMI ...   
7585     APP2135        Apple iPad Pro 12.9 "Wi-Fi 512GB Space Gray   
6882     APP2006  Watch Apple Nike + 38mm aluminum case in silve...   
10092  APP1973-A           Open - Apple iPad Wi-Fi 128GB Space Gray   
6384   APP1039-A                     Open - USB-C Apple USB Adapter   
6422     BEL0133       Belkin MIXIT Silver Lightning iPhone Support   
1620     MOS0141              Moshi iGlaze iPhone 6 / 6S Plus Black   
1703     QNA0081        QNAP TS-251 | 4GB RAM Mac and PC Server Nas   
8188     CRU0059       Crucial Mac Memory 16GB DDR4 2400MHz SO-DIMM   
5131     PAC1520  Pack QNAP TS-451A NAS 2GB RAM l l 24TB (4x6TB)...   
8179   TPL0032-A        Open - TP-LINK Cable 1m White Lightning IFM   
1605     PAC0758  Samsung SSD 850 expansion kit EVO 250GB iMac 2...   
3097     TUC0237          Housing Sottile Tucano iPhone 6 / 6S Gray   
5348     TUC0296          CHANGE Tucano iPhone 8/7 Transparent Case   
5606     BEA0067    Powerbeats3 Beats Wireless Headphones Red Siren   
8213     SYN0175   Synology M2D17 to M.2 PCIe Adapter Card SATA SSD   
9620   GTE0096-A  Open - G-Technology G-DRIVE HDD 10TB 7200rpm T...   
8959     PAC2455  DS218J Synology NAS Server | 4TB (2x2TB) Seaga...   

                                                    desc    price promo_price  \
3418   NAS server 4 bays and 4 GB RAM for small busin...   845.79   8.449.902   
153           SSD hard drive adapter or 2.5 for Mac Pro.    35.99     259.896   
6294   Aluminum adapter with USB Type-C 4K HDMI port ...    79.99     305.128   
7585                  New iPad Pro 12.9 inch Wi-Fi 512GB     1289  12.520.015   
6882   Apple Watch 38 mm GPS processor with dual core...      439   4.390.001   
10092    Apple iPad Wi-Fi 128GB Space Gray (MP2H2TY / A)   502.81      457.95   
6384      Reconditioned adapter USB-C to USB for MacBook       25     213.645   
6422   Loading dock + support synchronization with Li...    34.99     299.899   
1620   Rigid shell shock protection and rasguÌ ± os f...       35     249.901   
1703                    2-bay NAS server for Mac and PC.      398   2.998.997   
8188   Crucial memory Mac SO-DIMM 16GB 2400MHz DDR4 i...   211.99   2.119.896   
5131   NAS with 2GB of RAM and 4 HDDs 6TB for Mac and PC  1466.99  11.773.675   
8179   USB charging cable and Lightning 1 meter long ...    12.99      77.995   
1605   SSD upgrade kit 250GB iMac 27-inch Mid 2010 wi...   168.98   1.255.847   
3097                            Cover for iPhone 6 / 6S.      7.9      69.902   
5348   ultra thin and transparent with 15 cards of di...     12.9      89.903   
5606   Ac̼stico headphones and waterproof design for...   199.95   1.799.899   
8213      Adapter Card Dual M.2 SSD SSD cache to improve   148.99   1.489.897   
9620   Refurbished Hard Drive 10TB with 2 USB 3 and T...   683.99   4.463.934   
8959   2-bay NAS server entry-level to back domestic ...   405.97   3.371.786   

       in_stock       type  
3418          0   12175397  
153           1   12755395  
6294          0       1298  
7585          1   51861714  
6882          0   24885185  
10092         0  113851714  
6384          0   12585395  
6422          1   13615399  
1620          0   11865403  
1703          0   12175397  
8188          1       1364  
5131          1   12175397  
8179          0       1298  
1605          1       1433  
3097          0   11865403  
5348          1   11865403  
5606          0       5384  
8213          0       1404  
9620          0   11935397  
8959          0   12175397

In [64]:
# New column to count decimal numbers
s = products_df['price'].str.strip()
products_df['decimal_count'] = s.str.extract(r'\.(\d+)$')[0].str.len().fillna(0).astype(int)
products_df

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
10152  BEL0376        Belkin Travel Support Apple Watch Black   
10153  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
10154  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
10155  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
10156  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc  price promo_price  \
0           Aluminum support compatible with all MacBook  59.99     499.899   
1              USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                                 mouse Apple USB cable.     59     569.898   
3                  IPhone dock and USB Cable Apple iPod.     25     229.997   
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
...                                                  ...    ...         ...   
10152  compact and portable stand vertically or horiz...  29.99     269.903   
10153  Backpack with capacity of 14 liter compartment...  69.95     649.903   
10154  Backpack with capacity of 14 liter compartment...  69.95     649.903   
10155  Backpack with capacity of 14 liter compartment...  69.95     649.903   
10156  Backpack with capacity of 14 liter compartment...  69.95     649.903   

       in_stock      type  decimal_count  
0             1      8696              2  
1             0  13855401              0  
2             0      1387              0  
3             0      1230              0  
4             1      1364              2  
...         ...       ...            ...  
10152         1     12282              2  
10153         1      1392              2  
10154         1      1392              2  
10155         0      1392              2  
10156         1      1392              2  

[10157 rows x 8 columns]

In [65]:
# CHECK: Count the number of decimal places in the unit_price
products_df['decimal_count'].value_counts()

decimal_count
2    6271
0    3215
1     506
3     165
Name: count, dtype: int64

165 rows have more than 2 decimal places.

In [66]:
mask_3_decimal_count = products_df['decimal_count']==3

In [67]:
# Extract 'sku' of weird, 3-decimal count prices 
decimal_3_count_df =products_df.loc[mask_3_decimal_count,['sku','price']]
decimal_3_count_df

sku    price
201      REP0043  499.004
319      PIE0011   98.978
354      SEN0061  649.891
357      SEV0026  599.918
364      SEV0024  599.918
...          ...      ...
10022  FIB0002-A  599.918
10040    RIN0017  350.005
10108  TPL0030-A   54.329
10141    REP0428  599.906
10145    REP0429  599.906

[165 rows x 2 columns]

In [68]:
mask = orderlines_df['sku'].isin(decimal_3_count_df['sku'])

In [69]:
# Corresponding 'sku' in orderlines table
orderlines_df.loc[mask,:]

id  id_order  product_id  product_quantity        sku  \
38      1119186    299576           0                 1    SAM0095   
183     1119447    299695           0                 1    PAC1400   
366     1119800    299881           0                 1    SAN0137   
411     1119886    299925           0                 1    IFX0020   
567     1120198    300079           0                 1  KIN0149-A   
...         ...       ...         ...               ...        ...   
289411  1649659    527147           0                 1    SEV0024   
289446  1649720    527170           0                 1    SEV0024   
289462  1649748    527191           0                 1    FIB0007   
289662  1650146    527367           0                 1    REP0250   
289664  1650149    527367           0                 1    REP0299   

        unit_price                date  
38           82.64 2017-01-01 06:51:23  
183          90.99 2017-01-01 13:27:15  
366          49.39 2017-01-01 19:13:45  
411           7.99 2017-01-01 20:03:56  
567          59.03 2017-01-01 22:57:29  
...            ...                 ...  
289411       59.99 2018-03-14 12:05:02  
289446       59.99 2018-03-14 12:16:05  
289462       99.99 2018-03-14 12:19:19  
289662       69.90 2018-03-14 13:39:34  
289664       59.99 2018-03-14 13:40:04  

[3060 rows x 7 columns]

3060 orders are affected by prices with 3 decimal numbers (products_df). We will drop these orders to avoid erroneous prices in our analysis. We will also save the dropped id_orders to a new dataframe for reference.

In [70]:
# Before dropping the rows, we will save the dropped id_orders to a new dataframe
dropped_ids_3_decimal_sku_prices_df = pd.DataFrame(orderlines_df.loc[mask, 'id_order']).copy().reset_index(drop=True)
dropped_ids_3_decimal_sku_prices_df['reason'] = "Corrupted decimal places in column 'price' in products_df dataset" 
dropped_ids_3_decimal_sku_prices_df

id_order                                             reason
0       299576  Corrupted decimal places in column 'price' in ...
1       299695  Corrupted decimal places in column 'price' in ...
2       299881  Corrupted decimal places in column 'price' in ...
3       299925  Corrupted decimal places in column 'price' in ...
4       300079  Corrupted decimal places in column 'price' in ...
...        ...                                                ...
3055    527147  Corrupted decimal places in column 'price' in ...
3056    527170  Corrupted decimal places in column 'price' in ...
3057    527191  Corrupted decimal places in column 'price' in ...
3058    527367  Corrupted decimal places in column 'price' in ...
3059    527367  Corrupted decimal places in column 'price' in ...

[3060 rows x 2 columns]

In [71]:
# Drop the rows from the orderlines_df
orderlines_df = orderlines_df[~mask].reset_index(drop=True)
orderlines_df

id  id_order  product_id  product_quantity      sku  unit_price  \
0       1119109    299539           0                 1  OTT0133       18.99   
1       1119110    299540           0                 1  LGE0043      399.00   
2       1119111    299541           0                 1  PAR0071      474.05   
3       1119112    299542           0                 1  WDT0315       68.39   
4       1119113    299543           0                 1  JBL0104       23.74   
...         ...       ...         ...               ...      ...         ...   
286638  1650199    527398           0                 1  JBL0122       42.99   
286639  1650200    527399           0                 1  PAC0653      141.58   
286640  1650201    527400           0                 2  APP0698        9.99   
286641  1650202    527388           0                 1  BEZ0204       19.99   
286642  1650203    527401           0                 1  APP0927       13.99   

                      date  
0      2017-01-01 00:07:19  
1      2017-01-01 00:19:45  
2      2017-01-01 00:20:57  
3      2017-01-01 00:51:40  
4      2017-01-01 01:06:38  
...                    ...  
286638 2018-03-14 13:57:25  
286639 2018-03-14 13:57:34  
286640 2018-03-14 13:57:41  
286641 2018-03-14 13:58:01  
286642 2018-03-14 13:58:36  

[286643 rows x 7 columns]

In [72]:
# Dropping rows with three decimal numbers in the 'price' column
products_df = products_df[~mask_3_decimal_count].reset_index(drop=True)
products_df

sku                                           name  \
0     RAI0007              Silver Rain Design mStand Support   
1     APP0023              Apple Mac Keyboard Keypad Spanish   
2     APP0025               Mighty Mouse Apple Mouse for Mac   
3     APP0072  Apple Dock to USB Cable iPhone and iPod white   
4     KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...       ...                                            ...   
9987  BEL0376        Belkin Travel Support Apple Watch Black   
9988  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
9989  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
9990  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
9991  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                   desc  price promo_price  \
0          Aluminum support compatible with all MacBook  59.99     499.899   
1             USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                                mouse Apple USB cable.     59     569.898   
3                 IPhone dock and USB Cable Apple iPod.     25     229.997   
4     2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
...                                                 ...    ...         ...   
9987  compact and portable stand vertically or horiz...  29.99     269.903   
9988  Backpack with capacity of 14 liter compartment...  69.95     649.903   
9989  Backpack with capacity of 14 liter compartment...  69.95     649.903   
9990  Backpack with capacity of 14 liter compartment...  69.95     649.903   
9991  Backpack with capacity of 14 liter compartment...  69.95     649.903   

      in_stock      type  decimal_count  
0            1      8696              2  
1            0  13855401              0  
2            0      1387              0  
3            0      1230              0  
4            1      1364              2  
...        ...       ...            ...  
9987         1     12282              2  
9988         1      1392              2  
9989         1      1392              2  
9990         0      1392              2  
9991         1      1392              2  

[9992 rows x 8 columns]

In [73]:
# CHECK: Count the number of decimal points in the unit_price
products_df['price'].str.count("\\.").value_counts()

price
1    6777
0    3215
Name: count, dtype: int64

In [74]:
# CHECK: Count the number of decimal places in the unit_price
products_df['decimal_count'].value_counts()

decimal_count
2    6271
0    3215
1     506
Name: count, dtype: int64

In [75]:
# Removing 'decimal_count' column
products_df = products_df.drop(columns=['decimal_count'])

In [76]:
# Converting unit_price type to numeric
products_df["price"] = pd.to_numeric(products_df["price"])

Price column correctly converted to numeric!!!

In [77]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9992 entries, 0 to 9991
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sku          9992 non-null   object 
 1   name         9992 non-null   object 
 2   desc         9986 non-null   object 
 3   price        9992 non-null   float64
 4   promo_price  9992 non-null   object 
 5   in_stock     9992 non-null   int64  
 6   type         9946 non-null   object 
dtypes: float64(1), int64(1), object(5)
memory usage: 546.6+ KB


In [78]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 286643 entries, 0 to 286642
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                286643 non-null  int64         
 1   id_order          286643 non-null  int64         
 2   product_id        286643 non-null  int64         
 3   product_quantity  286643 non-null  int64         
 4   sku               286643 non-null  object        
 5   unit_price        286643 non-null  float64       
 6   date              286643 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(4), object(1)
memory usage: 15.3+ MB


In [79]:
# Concatenating drop id's dataframes

# Append to master log
dropped_ids_df = pd.concat([dropped_ids_missing_orders_total_paid_df, dropped_ids_missing_sku_prices_df, dropped_ids_mult_dots_sku_prices_df, dropped_ids_3_decimal_sku_prices_df], ignore_index=True)

dropped_ids_df

id_order                                             reason
0       427314   Missing column "total_paid" in orders_df dataset
1       427314   Missing column "total_paid" in orders_df dataset
2       431655   Missing column "total_paid" in orders_df dataset
3       447411   Missing column "total_paid" in orders_df dataset
4       448966   Missing column "total_paid" in orders_df dataset
...        ...                                                ...
7335    527147  Corrupted decimal places in column 'price' in ...
7336    527170  Corrupted decimal places in column 'price' in ...
7337    527191  Corrupted decimal places in column 'price' in ...
7338    527367  Corrupted decimal places in column 'price' in ...
7339    527367  Corrupted decimal places in column 'price' in ...

[7340 rows x 2 columns]

In [80]:
# Drop duplicates
dropped_ids_df = dropped_ids_df.drop_duplicates()
dropped_ids_df

id_order                                             reason
0       427314   Missing column "total_paid" in orders_df dataset
2       431655   Missing column "total_paid" in orders_df dataset
3       447411   Missing column "total_paid" in orders_df dataset
4       448966   Missing column "total_paid" in orders_df dataset
5       449596   Missing column "total_paid" in orders_df dataset
...        ...                                                ...
7334    526833  Corrupted decimal places in column 'price' in ...
7335    527147  Corrupted decimal places in column 'price' in ...
7336    527170  Corrupted decimal places in column 'price' in ...
7337    527191  Corrupted decimal places in column 'price' in ...
7338    527367  Corrupted decimal places in column 'price' in ...

[7252 rows x 2 columns]

In [81]:
# Find the percentage of dropped orders
percent_dropped = (100 * dropped_ids_df.shape[0] / (dropped_ids_df.shape[0] + orderlines_df.id_order.nunique()))
print(f"{percent_dropped:.2f}% of the orders in the orderlines DataFrame were dropped")

3.49% of the orders in the orderlines DataFrame were dropped


### Consistency Check:
Remove from 'orders_df' the orders that were dropped from orderlines_df.

In [82]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 226904 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      226904 non-null  int64         
 1   created_date  226904 non-null  datetime64[ns]
 2   total_paid    226904 non-null  float64       
 3   state         226904 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 8.7+ MB


In [83]:
mask = orders_df['order_id'].isin(dropped_ids_df['id_order'])

In [84]:
orders_df.loc[mask,:]

order_id        created_date  total_paid            state
9         246018 2017-02-13 23:02:57      211.95        Completed
15        251688 2017-11-24 23:57:07      505.76        Completed
54        265954 2017-10-02 15:18:15      127.02        Completed
84        278926 2017-11-23 11:06:39      795.85        Completed
235       299576 2017-01-01 06:51:23       82.64  Shopping Basket
...          ...                 ...         ...              ...
226698    527191 2018-03-14 12:19:19      368.50  Shopping Basket
226793    527286 2018-03-14 12:55:14     1165.99      Place Order
226849    527342 2018-03-14 13:24:51     1335.99      Place Order
226871    527364 2018-03-14 13:36:13     1329.00  Shopping Basket
226874    527367 2018-03-14 13:39:34      129.89      Place Order

[7123 rows x 4 columns]

7123 orders need to be removed from orders_df.

In [85]:
orders_df = orders_df[~mask].reset_index(drop=True)
orders_df

order_id        created_date  total_paid            state
0         241319 2017-01-02 13:35:40       44.99        Cancelled
1         241423 2017-11-06 13:10:02      136.15        Completed
2         242832 2017-12-31 17:40:03       15.76        Completed
3         243330 2017-02-16 10:59:38       84.98        Completed
4         243784 2017-11-24 13:35:19      157.86        Cancelled
...          ...                 ...         ...              ...
219776    527397 2018-03-14 13:56:38       42.99      Place Order
219777    527398 2018-03-14 13:57:25       42.99  Shopping Basket
219778    527399 2018-03-14 13:57:34      141.58  Shopping Basket
219779    527400 2018-03-14 13:57:41       19.98  Shopping Basket
219780    527401 2018-03-14 13:58:36       18.98      Place Order

[219781 rows x 4 columns]

About 3% of the orders were dropped from orders_df.

# Exporting Cleaned CSV Files

In [ ]:
orders_df.to_csv("3_Data_Cleaning/data/orders_df_cleaned.csv", index=False)
orderlines_df.to_csv("3_Data_Cleaning/data/orderlines_df_cleaned.csv", index=False)
products_df.to_csv("3_Data_Cleaning/data/products_df_cleaned.csv", index=False)


In [86]:
dropped_ids_df.to_csv("3_Data_Cleaning/data/dropped_ids_df.csv", index=False)

In [ ]:
from google.colab import files

#orders_df.to_csv("orders_df_cleaned.csv", index=False)
#files.download("orders_df_cleaned.csv")

#orderlines_df.to_csv("orderlines_df_cleaned.csv", index=False)
#files.download("orderlines_df_cleaned.csv")

#products_df.to_csv("products_df_cleaned.csv", index=False)
#files.download("products_df_cleaned.csv")